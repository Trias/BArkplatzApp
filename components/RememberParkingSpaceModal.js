import CloseableModal from './ClosableModal'
import renderTime from "./renderTime";
import React from 'react';
import {Text, TouchableOpacity} from 'react-native';
import TimePicker from 'react-native-24h-timepicker';

export class RememberParkingSpaceModal extends React.Component {
    state = {
        hourDuration: 1,
        minuteDuration: 0
    }

    render() {
        let { pressedParkingSpaceId, onRememberParkingSpace, ...rest } = this.props;
        return <CloseableModal
            title={`Parkplatz ${pressedParkingSpaceId} merken`}
            onAction={() => onRememberParkingSpace({
                parkingSpaceId: pressedParkingSpaceId,
                hourDuration: this.state.hourDuration,
                minuteDuration: this.state.minuteDuration,
            })}
            actionText={"Parkplatz merken"}
            {...rest}
        >
            <Text style={{ marginTop: 5, marginBottom: 15 }}>Wie lange wollen Sie parken?</Text>
            <TouchableOpacity
                style={styles.button}
                onPress={() => this.TimePicker.open()}
            >
                <Text style={styles.buttonText}>Zeit w√§hlen</Text>
            </TouchableOpacity>
            <Text>Erinnere mich in {renderTime(this.state.hourDuration, this.state.minuteDuration)} Stunden</Text>

            <TimePicker
                ref={(ref) => { this.TimePicker = ref }}
                minuteInterval={5}
                selectedHour={this.state.hourDuration + ""}
                selectedMinute={this.state.minuteDuration + ""}
                textCancel="Abbrechen"
                textConfirm="Fertig"
                onConfirm={(hour, minute) => { this.onRememberConfirm(hour, minute) }}
                onCancel={() => {
                    this.TimePicker.close();
                }} />
        </CloseableModal>
    }

    onRememberConfirm(hour, minute) {
        let hourDuration = parseInt(hour);
        let minuteDuration = parseInt(minute);
        this.setState({ hourDuration, minuteDuration });

        this.TimePicker.close();
    }
}

const styles = {
    button: {
        backgroundColor: "#4EB151",
        paddingVertical: 11,
        paddingHorizontal: 17,
        borderRadius: 3,
        marginVertical: 50
    },
    buttonText: {
        color: "#FFFFFF",
        fontSize: 16,
        fontWeight: "600"
    }
}