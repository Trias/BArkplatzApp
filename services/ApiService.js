const ACCESS_TOKEN = '';
const URL = ``;

import dummyData from '../data/UpdatedDummy.json';

let cache = null;

let interval = null;

async function rpcCall(method, params = {}) {
    let response;
    try {
        console.log('start API call: '+ method + ' with params: ' + JSON.stringify(params));
        response = await fetch(URL, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                method, params
            })
        });
        let json = await response.json()

        console.log('api call successful');
        console.log('method: ' + method);
        console.log('params: ' + JSON.stringify(params));
        console.log(json);

        return json;
    } catch (e) {
        console.warn('error while getting API data')
        console.log(response);
        console.log(e);
    }
}

export async function initApi() {
    await fetchAllAsync();
    interval = setInterval(fetchAllAsync, 60 * 1000);
}
export function stopApi() {
    clearInterval(interval);
    interval = null;
}
export async function fetchAllAsync() {
    let response = await rpcCall('fetchAll');
    if (response && !response.error) {
        cache = addCoordinates(response)
    } else {
        console.warn('empty data from api, using old data');
    }
    return cache;
}
export async function getReservationsOfThisUserAsync() {
    return await rpcCall('getReservationsOfThisUser')
}
export async function reserveAsync(reservation) {
    return await rpcCall('reserve', reservation)
}
export async function deleteReservationAsync(reservation) {
    return await rpcCall('deleteReservation', reservation);
}
export async function getReservationOfOneParkingSpaceAsync(parkingSpaceId) {
    return await rpcCall('getReservationOfOneParkingSpace', { parkingSpaceId });
}
export function getData() {
    if (!cache) {
        console.warn('error fetching data from api, using dummy data');
        cache = addCoordinates(dummyData);
    }

    return cache;
}

function addCoordinates(cache){
    for (let parkingLot of Object.values(cache)) {
        let firstParkingSpace = Object.values(parkingLot.spaces)[0];
        parkingLot.latitude = firstParkingSpace.latitude;
        parkingLot.longitude = firstParkingSpace.longitude;
    }

    return cache
}
