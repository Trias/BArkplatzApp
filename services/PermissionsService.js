import * as Permissions from 'expo-permissions';

let hasLocationPermission = false;
export async function requestLocationPermission() {
    if (hasLocationPermission) {
        return true;
    }
    try {
        const { status } = await Permissions.askAsync(Permissions.LOCATION);
        if (status === 'granted') {
            hasLocationPermission = true;
            console.log('gps granted');
            return true;
        } else {
            console.log('gps permission denied');
            return false;
        }
    } catch (err) {
        console.warn(err);
        return false;
    }
}

let hasNotificationPermission = false;
export async function requestNotificationPermission() {
    if (hasNotificationPermission) {
        return true;
    }
    try {
        const { status } = await Permissions.askAsync(Permissions.USER_FACING_NOTIFICATIONS);
        if (status === 'granted') {
            hasNotificationPermission = true;
            console.log('notifications granted');
            return true;
        } else {
            console.log('notifications denied');
            return false;
        }
    } catch (err) {
        console.warn(err);
        return false;
    }
}