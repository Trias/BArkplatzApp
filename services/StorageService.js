import { AsyncStorage } from "react-native";

const cache = {};

export async function initStorage(){
    const allKeys = await AsyncStorage.getAllKeys();
    for (let key of allKeys){
        let value = await AsyncStorage.getItem(key);
        cache[key] = value;
    }
}

export function getItem(name){
    return cache[name];
}

export function setItem(name, item){
    cache[name] = item;
    setItemAsync(name, item);
}

export function removeItem(name){
    delete cache[name];
    removeItemAsync(name);
}

export async function getItemAsync(name){
    return AsyncStorage.getItem(name)
}

export async function setItemAsync(name, item){
    cache[name] = item;
    return AsyncStorage.setItem(name, item);
}

export async function removeItemAsync(name){
    cache[name] = undefined;
    return AsyncStorage.removeItem(name);
}