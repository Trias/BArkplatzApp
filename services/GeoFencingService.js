import * as TaskManager from 'expo-task-manager';
import * as Location from 'expo-location';
import { Notifications } from 'expo';
import { requestLocationPermission, requestNotificationPermission } from './PermissionsService';

const TASK_NAME = "parkingLots";

let geoFencingActive = false;
export async function initGeofencing() {
  try {
    // permissions returns only for location permissions on iOS and under certain conditions, see Permissions.LOCATION
    let locationStatus = await requestLocationPermission();
    let notificationStatus = await requestNotificationPermission();

    if (!geoFencingActive) {
      if (locationStatus && notificationStatus) {
        geoFencingActive = true;
        await Location.startGeofencingAsync(TASK_NAME, [{
          // erba: 49.903, 10.87
          // harcoded for now to avoid dependency on network request
          identifier: "631", // heinrichsdamm
          latitude: 49.89,
          longitude: 10.903,
          radius: 100
        },
        {
          identifier: "38", // brennerstraÃŸe
          latitude: 49.901,
          longitude: 10.90,
          radius: 100
        }]);
      } else {
        console.log('Location permission or notification permission not granted');
      }
    }
  } catch (e) {
    console.log('error while geofencing');
  }
}

TaskManager.defineTask(TASK_NAME, ({ data: { eventType, region }, error }) => {
  console.log("geofencing event");
  if (error) {
    // check `error.message` for more details.
    console.log(error)
    return;
  }
  if (eventType === Location.GeofencingEventType.Enter) {
    const localNotification = {
      title: "Sie sind am Parkplatz angekommen!",
      body: "jetzt Parkplatz merken"
    }
    Notifications.addListener(notification => {
      if (notification.origin == "selected") {
        let { MainTabNavigator } = import('../navigation/MainTabNavigator');
        MainTabNavigator.navigation.navigate('ParkingLotDetail', { parkingLotId: region.identifier })
      }
    })
    Notifications.presentLocalNotificationAsync(localNotification);
    console.log("You've entered region:", region);
  } else if (eventType === Location.GeofencingEventType.Exit) {
    console.log("You've left region:", region);
  }
});