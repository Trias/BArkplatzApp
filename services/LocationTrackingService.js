import * as Location from 'expo-location';
import {requestLocationPermission} from './PermissionsService';

let locationTrackingInterval = null;
let locationTrackingActive = false;

export async function initLocationTracking(){
  try {
    // permissions returns only for location permissions on iOS and under certain conditions, see Permissions.LOCATION
    let locationStatus = await requestLocationPermission();
    if (!locationTrackingActive && locationStatus){
      locationTrackingActive = true;
      locationCache = await getLocationAsync();
      locationTrackingInterval = setInterval(getLocationAsync, 10000);
    }
  }catch(e){
    console.log('error while location');
  }
}

export function stopLocationTracking(){
  clearInterval(locationTrackingInterval);
  locationTrackingInterval = null;
  locationTrackingActive = false;
}

let locationCache = null;
export async function getLocationAsync(){
  locationCache = await Location.getCurrentPositionAsync({enableHighAccuracy: true});
  return locationCache;
}

export function getLocation(){
  if(locationCache){
    return locationCache;
  }else{
    console.warn('could not get location data, using fallback');
    return {coords: {latitude: 49.9028939, longitude: 10.8686975}};
  }
}