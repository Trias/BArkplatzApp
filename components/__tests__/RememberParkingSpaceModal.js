import React from 'react';
import renderer from 'react-test-renderer';

import {RememberParkingSpaceModal} from '../RememberParkingSpaceModal';

import * as ApiService from '../../services/ApiService';
import * as LocationTrackingService from '../../services/LocationTrackingService';

jest.mock('../../services/ApiService');
jest.mock('../../services/LocationTrackingService');

function addCoordinates(cache){
    for (let parkingLot of Object.values(cache)) {
        let firstParkingSpace = Object.values(parkingLot.spaces)[0];
        parkingLot.latitude = firstParkingSpace.latitude;
        parkingLot.longitude = firstParkingSpace.longitude;
    }

    return cache
}
ApiService.getData = jest.fn(() => {return addCoordinates(require('../../data/UpdatedDummy'))});

it(`renders correctly`, () => {
  const tree = renderer.create(<RememberParkingSpaceModal
  />).toJSON();

  expect(tree).toMatchSnapshot();
});
