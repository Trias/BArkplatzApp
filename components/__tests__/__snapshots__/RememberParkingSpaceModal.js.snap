// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Modal
  hardwareAccelerated={false}
  visible={true}
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "paddingBottom": 44,
        "paddingTop": 22,
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "lineHeight": 40,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 30,
            "zIndex": 100,
          }
        }
      >
        <Text />
      </View>
      <Text
        style={
          Object {
            "fontSize": 16,
            "fontWeight": "bold",
            "lineHeight": 44,
          }
        }
      >
        Parkplatz undefined merken
      </Text>
      <Text
        style={
          Object {
            "marginBottom": 15,
            "marginTop": 5,
          }
        }
      >
        Wie lange wollen Sie parken?
      </Text>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#4EB151",
            "borderRadius": 3,
            "marginVertical": 50,
            "opacity": 1,
            "paddingHorizontal": 17,
            "paddingVertical": 11,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 16,
              "fontWeight": "600",
            }
          }
        >
          Zeit w√§hlen
        </Text>
      </View>
      <Text>
        Erinnere mich in 
        01:00
         Stunden
      </Text>
      <Modal
        animationType="none"
        hardwareAccelerated={false}
        onRequestClose={[Function]}
        supportedOrientations={
          Array [
            "portrait",
            "portrait-upside-down",
            "landscape",
            "landscape-left",
            "landscape-right",
          ]
        }
        transparent={true}
        visible={false}
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#00000077",
                "flex": 1,
              },
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "transparent",
                "flex": 1,
                "opacity": 1,
              }
            }
          />
          <View
            onMoveShouldSetResponder={[Function]}
            onMoveShouldSetResponderCapture={[Function]}
            onResponderEnd={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderReject={[Function]}
            onResponderRelease={[Function]}
            onResponderStart={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            onStartShouldSetResponderCapture={[Function]}
            style={
              Object {
                "backgroundColor": "#fff",
                "height": 0,
                "overflow": "hidden",
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                ],
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "borderBottomWidth": 1,
                  "borderColor": "#ccc",
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "space-between",
                }
              }
            >
              <View
                accessible={true}
                isTVSelectable={true}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "opacity": 1,
                    "paddingHorizontal": 20,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#006BFF",
                        "fontSize": 20,
                        "fontWeight": "500",
                      },
                      Object {
                        "color": "#666",
                        "fontWeight": "400",
                      },
                    ]
                  }
                >
                  Abbrechen
                </Text>
              </View>
              <View
                accessible={true}
                isTVSelectable={true}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "opacity": 1,
                    "paddingHorizontal": 20,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#006BFF",
                      "fontSize": 20,
                      "fontWeight": "500",
                    }
                  }
                >
                  Fertig
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <RCTPicker
                  items={
                    Array [
                      Object {
                        "label": "0",
                        "textColor": undefined,
                        "value": "0",
                      },
                      Object {
                        "label": "1",
                        "textColor": undefined,
                        "value": "1",
                      },
                      Object {
                        "label": "2",
                        "textColor": undefined,
                        "value": "2",
                      },
                      Object {
                        "label": "3",
                        "textColor": undefined,
                        "value": "3",
                      },
                      Object {
                        "label": "4",
                        "textColor": undefined,
                        "value": "4",
                      },
                      Object {
                        "label": "5",
                        "textColor": undefined,
                        "value": "5",
                      },
                      Object {
                        "label": "6",
                        "textColor": undefined,
                        "value": "6",
                      },
                      Object {
                        "label": "7",
                        "textColor": undefined,
                        "value": "7",
                      },
                      Object {
                        "label": "8",
                        "textColor": undefined,
                        "value": "8",
                      },
                      Object {
                        "label": "9",
                        "textColor": undefined,
                        "value": "9",
                      },
                      Object {
                        "label": "10",
                        "textColor": undefined,
                        "value": "10",
                      },
                      Object {
                        "label": "11",
                        "textColor": undefined,
                        "value": "11",
                      },
                      Object {
                        "label": "12",
                        "textColor": undefined,
                        "value": "12",
                      },
                      Object {
                        "label": "13",
                        "textColor": undefined,
                        "value": "13",
                      },
                      Object {
                        "label": "14",
                        "textColor": undefined,
                        "value": "14",
                      },
                      Object {
                        "label": "15",
                        "textColor": undefined,
                        "value": "15",
                      },
                      Object {
                        "label": "16",
                        "textColor": undefined,
                        "value": "16",
                      },
                      Object {
                        "label": "17",
                        "textColor": undefined,
                        "value": "17",
                      },
                      Object {
                        "label": "18",
                        "textColor": undefined,
                        "value": "18",
                      },
                      Object {
                        "label": "19",
                        "textColor": undefined,
                        "value": "19",
                      },
                      Object {
                        "label": "20",
                        "textColor": undefined,
                        "value": "20",
                      },
                      Object {
                        "label": "21",
                        "textColor": undefined,
                        "value": "21",
                      },
                      Object {
                        "label": "22",
                        "textColor": undefined,
                        "value": "22",
                      },
                      Object {
                        "label": "23",
                        "textColor": undefined,
                        "value": "23",
                      },
                    ]
                  }
                  onChange={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  selectedIndex={1}
                  style={
                    Array [
                      Object {
                        "height": 216,
                      },
                      Object {},
                    ]
                  }
                />
              </View>
              <Text
                style={
                  Object {
                    "alignSelf": "center",
                    "fontSize": 16,
                  }
                }
              >
                :
              </Text>
              <View
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <RCTPicker
                  items={
                    Array [
                      Object {
                        "label": "00",
                        "textColor": undefined,
                        "value": "00",
                      },
                      Object {
                        "label": "05",
                        "textColor": undefined,
                        "value": "05",
                      },
                      Object {
                        "label": "10",
                        "textColor": undefined,
                        "value": "10",
                      },
                      Object {
                        "label": "15",
                        "textColor": undefined,
                        "value": "15",
                      },
                      Object {
                        "label": "20",
                        "textColor": undefined,
                        "value": "20",
                      },
                      Object {
                        "label": "25",
                        "textColor": undefined,
                        "value": "25",
                      },
                      Object {
                        "label": "30",
                        "textColor": undefined,
                        "value": "30",
                      },
                      Object {
                        "label": "35",
                        "textColor": undefined,
                        "value": "35",
                      },
                      Object {
                        "label": "40",
                        "textColor": undefined,
                        "value": "40",
                      },
                      Object {
                        "label": "45",
                        "textColor": undefined,
                        "value": "45",
                      },
                      Object {
                        "label": "50",
                        "textColor": undefined,
                        "value": "50",
                      },
                      Object {
                        "label": "55",
                        "textColor": undefined,
                        "value": "55",
                      },
                    ]
                  }
                  onChange={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  selectedIndex={0}
                  style={
                    Array [
                      Object {
                        "height": 216,
                      },
                      Object {},
                    ]
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </Modal>
      <View
        style={
          Object {
            "marginTop": 20,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityStates={Array []}
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "padding": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Parkplatz merken
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</Modal>
`;
